Technical and editorial changes from review the EDID to Rx Caps mapping (#41)
AMWA-TV/nmos-edid-connection-management: docs/2.0. EDID to Receiver Caps mapping.md

# AMWA BCP-XXX-YY: EDID to NMOS Receiver Capabilities Mapping \[Work In Progress\]

{:.no_toc}

* A markdown unordered list which will be replaced with the ToC, excluding the "Contents header" from above

{:toc}

> Extended Display Identification Data (EDID) is a metadata format for an apparatus to describe its capabilities as a video source (e.g. graphics card or set-top box). The data format is defined by a standard published by the Video Electronics Standards Association (VESA). This document is targeted against [E-EDID A2][E-EDID] which consist of EDID 1.4 (and covers EDID 1.3) and is refered to as _Base EDID_ and the [CTA-861-G][CTA-861] Extension Block imposed by HDMI. The information present in the EDID is subject to the requirements of the [Display Monitor Timing (DMT)][DMT] specification Version 1 Rev 12.

Reviewer: > The information present in the EDID is subject to the requirements of the [Display Monitor Timing (DMT)][DMT] specification Version 1 Rev 12.

It follows from this sentence that DMT is a higher-level restriction for E-EDID. Will it be more precise to write that any DMT mentions in the text below refer DMT v1r12?

Author: The version is here and in the normative references which is always pointed at, on the hover of the link it shows the version. 🤔 

I think like this is fine

## Scope

This document provides Best Current Practice guidelines for how to implement a mapping of EDID fields to Receiver Capabilities in order to have a specific way of representing an EDID sink devices through an associated [IS-04][IS-04] Receiver.

## Use of Normative Language

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119][RFC-2119].

## Normative References

These appear at the end of the Markdown source for this document, and are referenced as hyperlinks within the main body.

## Definitions

_See also the [NMOS Glossary](https://specs.amwa.tv/nmos/branches/main/docs/4.0._Glossary.html), and definitions within RFCs._

### Gateway Device

A device which encapsulates a digital video or/and audio signal (usually, from a physical connection) into IP packets. Conversely, the device MAY perform the reverse operation for IP packets to a digital signal.

### Sink

A media consuming unit described by an EDID which can be associated to a Receiver.

## Introduction

> Gateway Devices which bridge a NMOS Receivers to baseband connection need to expose the capabilities of the baseband interface through the NMOS resouce to enable connection management. This can be obtained by mapping a Sink's EDID to a Receiver's Capabilities and factoring the abilities of the Gateway Device.

Author: ```suggestion
Gateway Devices which bridge a NMOS Receivers to baseband connection need to expose the capabilities of the baseband interface through the NMOS resource to enable connection management. This can be obtained by mapping a Sink's EDID to a Receiver's Capabilities and factoring the abilities of the Gateway Device.
```

The mapping is affected when a Gateway Device is:
- capable of altering a signal (e.g. it is able to consume 4K and downscale it for the Sink who supports up to HD resolutions)
- constrained beyond the capabilities of the Sink (e.g. it suports resolution up to 1080 but is connected to a 4k monitor)
- using any internal processing before pasing the signal to the Sink (e.g. support multiple Sinks per Receiver and use greatest common denominator to constrain itself) 

> The Gatewat Device MUST include any types of conversion, omit any restrictions, or reflect the internal processing in the capabilities of the Receiver.

Reviewer: ```suggestion
The Gateway Device MUST include any types of conversion, omit any restrictions, or reflect the internal processing in the capabilities of the Receiver.
```

> The Gateway Device MUST include any types of conversion, omit any restrictions, or reflect the internal processing in the capabilities of the Receiver.

Author: ```suggestion
The Gateway Device SHALL include any types of conversion, omit any restrictions, or reflect the internal processing in the capabilities of the Receiver.
```

## Video Receivers

If an [IS-04][IS-04] Video Receiver is associated with a Sink which has EDID, its supported video formats MUST be mapped into Receiver's Capabilities according to the rules below.

### Video timings

Video timings, described in [E-EDID][E-EDID], provide information about video frame size and rate. There are multiple blocks which keep information about timings each with their own mapping requirements.

Each video timing SHOULD be present in the Receiver's Capabilities as a Constraint Set with a non-empty

* `urn:x-nmos:cap:format:frame_width`
* `urn:x-nmos:cap:format:frame_height`

> * `urn:x-nmos:cap:format:grain_rate` which MUST be the frame rate of the signal

Author: ```suggestion
* `urn:x-nmos:cap:format:grain_rate` which MUST be the exact frame rate of the signal
```

      "enum": [ 768 ]
    },
    "urn:x-nmos:cap:format:interlace_mode": {
      "enum": [ "progressive" ]
    },
    "urn:x-nmos:cap:format:grain_rate": {
      "enum": [ { "numerator": 60 } ]
    }
  }
]
```

</details>

#### Standard Timings

Standard Timing format is defined in [E-EDID][E-EDID] section 3.9. The mapping is as follows:

* `urn:x-nmos:cap:format:frame_width` MUST be calculated from the _Horizontal Active Pixel Count_
* `urn:x-nmos:cap:format:frame_height` MUST be calculated using the _Frame Width_ and _Image Aspect Ratio_
* `urn:x-nmos:cap:format:grain_rate` MUST be calculated with _Field Refresh Rate_

<details><summary>Example: Standard Timings</summary>

For example, given the bytes of Standard Timings were `D1 C0 01 01 01 01 01 01 01 01 01 01 01` then the receiver capabilies could contain the following in its constraint sets

```json
[
  {
    "urn:x-nmos:cap:format:frame_width": {
      "enum": [ 1080 ]
    },
    "urn:x-nmos:cap:format:frame_height": {
      "enum": [ 1920 ]
    },
    "urn:x-nmos:cap:format:grain_rate": {
      "enum": [ { "numerator": 60 } ]
    }
  }
]
```

</details>

#### Detailed Timing Descriptors (18 Byte Descriptors)

> Defined in [E-EDID][E-EDID] section 3.10, there a 4 possible descriptors which can be provided. The first, _Prefered Timing Mode_ is an obligation and MUST have a `urn:x-nmos:cap:meta:preference` value off `100`.

Reviewer: ```suggestion
Defined in [E-EDID][E-EDID] section 3.10, there a 4 possible descriptors which can be provided. The first, _Prefered Timing Mode_ is an obligation and MUST have a `urn:x-nmos:cap:meta:preference` value of `100`.
```

> Defined in [E-EDID][E-EDID] section 3.10, there a 4 possible descriptors which can be provided. The first, _Prefered Timing Mode_ is an obligation and MUST have a `urn:x-nmos:cap:meta:preference` value of `100`. Any additional Detailed Timing Descriptors in [CTA-861][CTA-861] Extension BLock (section 7.2.1) MUST follow the mapping.

Reviewer: ```suggestion
Defined in [E-EDID][E-EDID] section 3.10, there are 4 possible descriptors which can be provided. The first, _Prefered Timing Mode_ is an obligation and MUST have a `urn:x-nmos:cap:meta:preference` value of `100`. Any additional Detailed Timing Descriptors in [CTA-861][CTA-861] Extension BLock (section 7.2.1) MUST follow the mapping.
```
What does "follow the mapping" mean?

Author: I stole the words from the next paragraph

![image](https://user-images.githubusercontent.com/16867443/114758548-15cf6500-9d2b-11eb-9950-ed2168f10f51.png)

Reviewer: This sentence makes me feeling like _Preferred Timing Mode_ should follow some other mapping because of "additional". Do you mind removing this word?

> Defined in [E-EDID][E-EDID] section 3.10, there are 4 possible descriptors which can be provided. The first, _Prefered Timing Mode_ is an obligation and MUST have a `urn:x-nmos:cap:meta:preference` value of `100`. Any additional Detailed Timing Descriptors in [CTA-861][CTA-861] Extension BLock (section 7.2.1) MUST follow the mapping.

Author: ```suggestion
Defined in [E-EDID][E-EDID] section 3.10, there are 4 possible descriptors which can be provided. The first, _Prefered Timing Mode_ is an obligation and MUST have a `urn:x-nmos:cap:meta:preference` value of `100`. Any Detailed Timing Descriptors in [CTA-861][CTA-861] Extension BLock (section 7.2.1) MUST follow the mapping.
```

* `urn:x-nmos:cap:format:frame_height` MUST be calculated using the _Vertical Addressable Video in lines_ which MUST be multiplied by 2 when _Signal Interface Type_ indicates interlaced
* `urn:x-nmos:cap:format:grain_rate` SHALL be accurately represented by
  * `numerator` MUST be calculated using _Pixel Clock_
  * `denominator` MUST be calculated with _Horizontal Addressable Video in pixels_, _Horizontal Blanking in pixels_, _Vertical Addressable Video in lines_ and _Vertical Blanking in lines_

#### 3 Byte CVT Codes

3 Byte CVT Code structure is defined in [E-EDID][E-EDID] section 3.10.3.8.

* `urn:x-nmos:cap:format:frame_height` MUST be calculated using value of the CVT Code
* `urn:x-nmos:cap:format:frame_width` MUST be calculated with Frame Height and _Aspect Ratio_

_Preferred Vertical Rate_ united with the frame width and height makes a Constraint Set and SHOULD have a `urn:x-nmos:cap:meta:preference` value off `50`

Each _Supported Vertical Rate_, except _Preferred Vertical Rate_, MUST be extracted to a new Constraint Set with the same frame width and height.

#### DMT Standard Codes & IDs Summary

> DMT Code are used to augement the Standard Timining Descriptors and CTV Codes, for some EDID IDs extra calcuations are REQUIRED to obtain the frame rate as per [DMT][DMT] section 2 and section 4.

Reviewer: ```suggestion
DMT Codes are used to augment the Standard Timing Descriptors and CVT Codes, for some EDID IDs extra calcuations are REQUIRED to obtain the frame rate as per [DMT][DMT] section 2 and section 4.
```

> * `urn:x-nmos:cap:format:grain_rate` MUST be calcuated and represent tge exact frame rate. The rational MAY be reduced by its greatest common denominator.

Reviewer: ```suggestion
* `urn:x-nmos:cap:format:grain_rate` MUST be calculated and represent the exact frame rate. The rational MAY be reduced by its greatest common denominator.
```

`YCbCr 4:2:0` Capability Map Data Block ([CTA-861][CTA-861] section 7.5.11) shows which timings support `YCbCr 4:2:0` in addition to subsampling formats listed in CTA Extension Header.

These values MUST be transformed into `urn:x-nmos:cap:format:color_sampling` with `enum` type value and added to each Constraint Set.

`YCbCr 4:2:0` Video Data Block ([CTA-861][CTA-861] section 7.5.10) marks timings as supporting only `YCbCr 4:2:0`.

Constraint Set associated with these timings MUST have `urn:x-nmos:cap:format:color_sampling` equal to `YCbCr-4:2:0`.

## Audio Receivers

If Basic Audio support bit is active in CTA Extension Header, audio receiver MUST have Receiver Capabilities.

When no descriptors are provided, the cabilities MUST contain:

* `urn:x-nmos:cap:format:media_type` MUST be equal to `audio/L8`
* `urn:x-nmos:cap:format:channel_count` MUST be equal to 2

### Short Audio Descriptors

If there are Short Audio Descriptors (see [CTA-861][CTA-861] section 7.5.2), then each of them MUST be transformed into a Constraint Set.

* `urn:x-nmos:cap:format:media_type` MUST be determined with _Audio Format Code_

>   * Linear PCM bit depth MUST be the third Short Audio Descriptors byte

Reviewer: ```suggestion
  * Linear PCM bit depth MUST be taken from the third Short Audio Descriptors byte
```
